### PR TITLE
Parse canonical tests v6

### PR DESCRIPTION
**Pull request by agarciamontoro**
I was sad to see that `cooklang-go` didn't have the green tick for passing canonical tests in https://cooklang.org/docs/for-developers/, so I'm trying to get them to work.

This is the first step towards that: updating the canonical tests to v6 and making them all parse (I just needed to accept single word timers with no duration nor unit).

Let me know if this makes sense, I'll continue working in the missing comparison step in the meantime.

Thanks for this library!

### CODE REVIEW

**Comment by aquilax** [comment]
Hi @agarciamontoro . Thank you for your contribution.

The library is somehow incompatible with the spec, as it produces segmented structure instead of sequential list of the elements but that was good enough at the time I worked on it. 

The spec test currently only checks if the parser does not fail but does not compare the actual results, which I would guess the check marks are for on the page.

**Comment by agarciamontoro** [comment]
Thanks, @aquilax!

> The library is somehow incompatible with the spec, as it produces segmented structure instead of sequential list of the elements but that was good enough at the time I worked on it.

Yes, I've noticed that. I've been thinking about it and I think that a good solution would be adding indexes to ingredients/timers/cookware so that we know where in the clean string they start. This way, we can split them afterwards if needed (for matching the spec tests or for whatever other reason, like highlighting them when formatting in e.g. html). We could also mimic what the spec does literally and slice the steps in intertwined directions/ingredient/directions/timer/..., but I think that's a worse solution, since it introduces some more complexity.

TL;DR: My idea is to convert this:

```go
Step{
	Directions: "Make 6 pizza balls using tipo zero flour, water, salt and fresh yeast. Put in a fridge for 2 days.",
	Timers:     []Timer{{Duration: 2, Unit: "days"}},
	Ingredients: []Ingredient{
		{Name: "tipo zero flour", Amount: IngredientAmount{true, 820., "820", "g"}},
		{Name: "water", Amount: IngredientAmount{true, 533, "533", "ml"}},
		{Name: "salt", Amount: IngredientAmount{true, 24.6, "24.6", "g"}},
		{Name: "fresh yeast", Amount: IngredientAmount{true, 1.6, "1.6", "g"}},
	},
	Cookware: []Cookware{{Name: "fridge", Quantity: 1, IsNumeric: false, QuantityRaw: ""}},
}
```

into this:

```go
Step{
	Directions: "Make 6 pizza balls using tipo zero flour, water, salt and fresh yeast. Put in a fridge for 2 days.",
	Timers:     []Timer{{Duration: 2, Unit: "days", Idx: 91}},
	Ingredients: []Ingredient{
		{Name: "tipo zero flour", Amount: IngredientAmount{true, 820., "820", "g"}, Idx: 25},
		{Name: "water", Amount: IngredientAmount{true, 533, "533", "ml"}, Idx: 42},
		{Name: "salt", Amount: IngredientAmount{true, 24.6, "24.6", "g"}, Idx: 49},
		{Name: "fresh yeast", Amount: IngredientAmount{true, 1.6, "1.6", "g"}, Idx: 58},
	},
	Cookware: []Cookware{{Name: "fridge", Quantity: 1, IsNumeric: false, QuantityRaw: "", Idx: 80}},
}
```

Where the new `Idx` show  where in the `Directions` string the ingredient/cookware/timer starts.

> The spec test currently only checks if the parser does not fail but does not compare the actual results, which I would guess the check marks are for on the page.

I plan on working on it so that the library can match the results in all tests :) If I implement the solution above, would you be open to merge it?

**Comment by aquilax** [comment]
@agarciamontoro

It's not a bad idea to include the indexes, though there are some caveats:
- you still need to know the length of the replaced segment in addition to the index
- sequential processing gets tricky from start to end, but can be done from end to start, and then need to reverse, which adds extra cycles.
- I'd use `Index` instead of `Idx` to make it more clear
- I'm not sure if there is a real world use case for the reconstruction of the directions other than passing the spec

That being said, I am not averse to merging a solution like this as it's pretty much backwards compatible with the current API.

**Comment by agarciamontoro** [comment]
> you still need to know the length of the replaced segment in addition to the index

We already know this with the `Name` in case of ingredients/cookware, and reconstructing the formatted string (`fmt.Sprintf("%v %s", timer.Duration, timer.Unit)`) for the timers.

> sequential processing gets tricky from start to end, but can be done from end to start, and then need to reverse, which adds extra cycles.

Not sure what you mean here, but I went for the dumbest solution: just compute the indices once the `Directions` string is already built, finding their names again in the string.

> I'd use Index instead of Idx to make it more clear

That makes sense :)

> I'm not sure if there is a real world use case for the reconstruction of the directions other than passing the spec

Exactly, that's why I thought the `Idx` solution was the best one, as it's the less intrusive.

I have [a branch](https://github.com/aquilax/cooklang-go/compare/master...agarciamontoro:cooklang-go:pass.canonical.tests.v6) (not ready for review yet, but in case you want to take a look), where I've already implemented the computation of those `Idx` and the tests are passing, so it does work. I haven't finished the conversion `cooklang.Recipe -> Result` yet, but I'm on it.

I'm ok opening a PR when that's ready and discussing improvements/fixes/optimizations :)

**Comment by aquilax** [comment]
I am thinking of changing the parser a bit and ultimately supporting the existing format in addition to the one used in the specs. Still WIP but you can take a peek here https://github.com/aquilax/cooklang-go/commit/94f3831afd8cb7a2f2d6c283e6e1c61fc45f72eb

**Comment by agarciamontoro** [comment]
Ah, that's neat! I'll try to finish what I started last week this weekend, let's see if life lets me. But if you're supporting the new format directly, that's a better solution.
